### PR TITLE
Fix the IDE freezing when an error is shown. 

### LINKED ISSUE #9: Waiting for OpenAI dialog hangs when insufficient_quota error occurs
Feedback od Šaňa:

Kolega pouzival (docgen v Pycharmu na Windows) a pri jeho free GPT verzi mu to vyhodilo tuto chybu pri generovani docstring, coz je asi ocakavane:

image

Problem je, ze pak ten window, ktory akoby caka na odpoved openAI sa zasekne a neda sa zavriet ani to Cancel tlacitko nereaguje. Pak je nutny restart PyCharmu.

### PR DESCRIPTION
**Pull request by msvana**
I also noticed that the Cancel button didn't work because the OpenAI API request was synchronous. When the request was running it wasn't possible to to check if the user clicked the Cancel button. To fix this I made the request asynchronous.

### CODE REVIEW

File: src/main/java/com/profiq/codexor/GenerateDocumentationAction.java [hunk 1]
```diff
@@ -63,20 +67,43 @@ public void actionPerformed(@NotNull AnActionEvent e) {
...
+
+                    responseFuture.exceptionally(e -> {
+                        if (e instanceof IOException || e instanceof InterruptedException) {
+                            showError("Can't reach OpenAI at the moment");
+                        }
+                        return null;
+                    });
+
+
+                    while (!responseFuture.isDone()) {
```
Comment (msvana) [inline src/main/java/com/profiq/codexor/GenerateDocumentationAction.java:93]: I wonder if there is some other way of checking whether the user clicked the Cancel button than a while loop. But it works.

**Comment by petrvecera** [comment]
Preferably we should not do a long task in the UI thread. 
https://plugins.jetbrains.com/docs/intellij/general-threading-rules.html

We should spawn a new thread and do a request in it. However I don't think it's worth investing our time right now. 
We can log this as an ticket for improvement 👍 

FYI ChatGPT:
When developing plugins for IntelliJ or any other JetBrains IDE, it's essential to ensure that long-running operations, especially IO-bound tasks like network requests, don't run on the Event Dispatch Thread (EDT). Running such tasks on the EDT would freeze the UI, leading to a poor user experience.

To avoid this, you can use IntelliJ's built-in concurrency utilities. Here are the steps to perform a network request without freezing the UI:

1. **Background Task**: Use `ProgressManager` to run the task in the background.

```java
ProgressManager.getInstance().run(new Task.Backgroundable(project, "Title of your task", false) {
    @Override
    public void run(@NotNull ProgressIndicator indicator) {
        // Your network request code goes here
    }
});
```

2. **Updating UI After Task Completion**: If you want to update the UI after your background task completes, ensure you do so on the EDT using `ApplicationManager.getApplication().invokeLater()`:

```java
ProgressManager.getInstance().run(new Task.Backgroundable(project, "Title of your task", false) {
    @Override
    public void run(@NotNull ProgressIndicator indicator) {
        // Your network request code
        
        String result = ...; // Let's assume you got some result from the request

        ApplicationManager.getApplication().invokeLater(() -> {
            // Update UI with the result
        });
    }
});
```

3. **Show Progress**: If you want to show the progress to the user, you can utilize the `ProgressIndicator`:

```java
@Override
public void run(@NotNull ProgressIndicator indicator) {
    indicator.setIndeterminate(false);
    indicator.setText("Fetching data...");

    for (int i = 0; i <= 100; i++) {
        if (indicator.isCanceled()) {
            // Handle cancellation
            return;
        }

        // Simulate network task or some processing
        try {
            Thread.sleep(50);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }

        indicator.setFraction(i / 100.0);
    }
}
```

4. **Handle Cancellations**: It's also a good idea to periodically check if the task was cancelled by the user, using the `ProgressIndicator.isCanceled()` method, and handle it accordingly.

5. **Network Request Libraries**: Remember to use a good networking library that supports cancellation, timeouts, etc. Libraries like OkHttp are great choices for this purpose.

By following these steps, you can ensure that your IntelliJ plugin performs network operations without blocking the UI, offering a responsive and user-friendly experience.

**Comment by msvana** [comment]
Well I am essentially doing the same thing. But instead of `Task.Backgroundable` I use `Task.Modal` so there is a progress bar in the middle of the screen. `TaskBackgroundable` would most likely just show a progress indicator in the status bar at the bottom (just like when WebStorm is indexing your project or something like that). 

I was essentially interested in replacing the `while` loop with something else. But ChatGPT suggested a `for` loop instead. So basically the same solution.

**Review by petrvecera — APPROVED** [review]
Let's ship it :shipit: 🚀

**Pull request merged**